docs: Add confidential-smart-contract.md (#989)
oasisprotocol/oasis-sdk: docs/contract/confidential-smart-contract.md

# Confidential Smart Contract

In this chapter we are going to see how to:
1. write a smart contract which stores and loads data to and from a
   confidential store and
2. instantiate and call the smart contract without revealing the call arguments.

> ## PublicCell VS ConfidentialCell

Reviewer: ```suggestion
## PublicCell vs ConfidentialCell
```

In the [hello world](./hello-world.md) example we used

> `oasis_contract_sdk_storage::cell::PublicCell<T>` to access the key-value store

Reviewer: Could we simplify that to just `PublicCell<T>`?

of that contract instance. In this case the value was stored unencrypted on the
blockchain associated with the hash of the key we provided in the constructor.

Cipher supports another primitive

> `oasis_contract_sdk_storage::cell::ConfidentialCell<T>` which allows you to

Reviewer: Same here.

> access the data of the contract instance confidentially assured by a

Reviewer: ```suggestion
access the data of the contract instance confidentially assured by
```

hardware-level encryption.

In contrast to non-confidential calls the hash of the key we passed to the
constructor, the runtime library also computes a nonce each time the data is
accessed. This nonce ensures the confidentiality so that the accessed key
appears different each time from the blockchain observer and even from the node

> operator. It is generated from:

Reviewer: This is a bit confusing. So the ever-changing nonce is for the values, not the keys as the keys need to be located at well-known locations. The keys are constant, but the value encryption uses the described nonce scheme.

- the round number,
- the number of the call inside the current smart contract execution,

> - the number of the smart contract executions until the contract was
>   instantiated.

Reviewer: ```suggestion
- the number of confidential storage accesses from smart contracts in the current block.
```

> As an additional safety measure, you will also need to provide the

Reviewer: This is not a safetly measure, just that you need to somehow pass the store instance.

instance of the *confidential store* of the contract instance. This is obtained
by calling `confidential_store()` of the contract's *context* object. If, for
example, the node operator tries to execute your code in a non-confidential

> environment, this method will panic.

Reviewer: It would be better to say that in a non-confidential environment the node operator will not be able to obtain the keys required to perform decryption so the operation would fail.

extern crate alloc;

use oasis_contract_sdk as sdk;
use oasis_contract_sdk_storage::cell::ConfidentialCell;

/// All possible errors that can be returned by the contract.
///
/// Each error is a triplet of (module, code, message) which allows it to be both easily
/// human readable and also identifyable programmatically.
#[derive(Debug, thiserror::Error, sdk::Error)]
pub enum Error {
    #[error("bad request")]
    #[sdk_error(code = 1)]
    BadRequest,
}

/// All possible requests that the contract can handle.
///
/// This includes both calls and queries.
#[derive(Clone, Debug, cbor::Encode, cbor::Decode)]
pub enum Request {
    #[cbor(rename = "instantiate")]
    Instantiate { initial_counter: u64 },

    #[cbor(rename = "say_hello")]
    SayHello { who: String },
}

/// All possible responses that the contract can return.
///
/// This includes both calls and queries.
#[derive(Clone, Debug, PartialEq, cbor::Encode, cbor::Decode)]
pub enum Response {
    #[cbor(rename = "hello")]
    Hello { greeting: String },

    #[cbor(rename = "empty")]
    Empty,
}

/// The contract type.
pub struct HelloWorld;

/// Storage cell for the counter.

> const COUNTER: ConfidentialCell<u64> = ConfidentialCell::new(b"counter");

Reviewer: A better example would be one where counters are scoped per-caller, so I can't read your counter value. Or something.

You can use a `ConfidentialMap<Address, u64>` for this.

But of course this is slightly more complicated.

Author: I agree, but let's merge the current version first and we'll extend it later. Current TODO in my head
- add `oasis contracts upgrade` to hello-world.md
- add usage of instantiate, upgrade policies to hello-world.md
- add `cargo test` in hello-world.md, mention x86-only for now, try using docker on mac?
- add `ConfidentialMap<K, V>` to confidential-smart-contract.md
- mention that keys are still non-confidential and add `ConfidentialCell<BTreeMap<K,V>>` to make keys confidential too, but for more gas.

Reviewer: Sounds good!

> revamp https://github.com/oasisprotocol/oasis-sdk/tree/main/tests/contracts (rename to example-contracts, split to hello-world, confidential-hello-world (do we have abbreviation for "confidential", c10l?), start using code snippets in the docs)

Note that the ones under `tests/contracts` are actually used in E2E tests so they shouldn't be changed. But yes, we can introduce additional "example contracts" as well and start using snippets.

            }
        );

        // Dispatch another SayHello message.
        let rsp = HelloWorld::call(
            &mut ctx,
            Request::SayHello {
                who: "second call".to_string(),
            },
        )
        .expect("SayHello call should work");

        // Make sure the greeting is correct.
        assert_eq!(
            rsp,
            Response::Hello {
                greeting: "hello second call (12)".to_string()
            }
        );
    }
}
```

The contract is built the same way as its non-confidential counterpart:

```shell
cargo build --target wasm32-unknown-unknown --release
```

::: warning

The blockchain store containing all compiled contracts is public. This means
that anyone will be able to decompile your smart contract and see how it
works. **Do not put any sensitive data inside the smart contract code!** 

:::

Since the smart contracts store is public, uploading the WASM code is
the same as for the non-confidential ones:

```shell
oasis contracts upload hello_world.wasm
```

## Confidential Instantiation and Calling

To generate a confidential transaction, the `oasis contracts` subcommand

> accepts `--encrypted` flag. Confidential transactions have encrypted contract

Reviewer: ```suggestion
accepts an `--encrypted` flag. Confidential transactions have encrypted contract
```

address, function name, parameters and the amounts and types of tokens sent.
**However, the _authorization information_ which contains information on the
signer is public!** Usually this field contains the public key of your account

> or a list of expected multisig keys.

Reviewer: Together with the gas limit and amount of fee to be paid for processing the transaction.

::: danger

While the transaction itself is confidential, the effects of a smart contract
execution may reveal some information. For example, the account balances are
public. If the effect is, say, subtraction of 10 tokens from the signer's
account, this most probably implies that they have been transferred as part of
this transaction.

:::

To instantiate a smart contract with the code ID obtained above, you can run:

```shell
oasis contracts instantiate CODEID '{instantiate: {initial_counter: 42}}' --encrypted
```

As is the case with non-confidential contracts, we obtain an incremented
instance ID. Finally, we make a confidential call:

```shell
oasis contracts call INSTANCEID '{say_hello: {who: "me"}}' --encrypted
```

::: danger

Regardless of the confidential storage used in the smart contract, any [emitted
event][emit_event] will be public.

:::

[emit_event]:

>   https://github.com/oasisprotocol/oasis-sdk/blob/main/runtime-sdk/src/context.rs#L804-L808

Reviewer: Since we are in a contract tutorial it would be better to link to the [contract-sdk context](https://api.docs.oasis.dev/oasis-sdk/oasis_contract_sdk/context/trait.Context.html#tymethod.emit_event). Also note that we can link to our published Rustdoc.

Author: Yeah, I also added rustdoc links to PublicCell and ConfidentialCell above.
